File: AliyunOSSiOSTests/OSSCancellationTests.m
//
//  OSSCancellationTests.m
//  AliyunOSSiOSTests
//
//  Created by 怀叙 on 2017/11/15.
//  Copyright © 2017年 zhouzhuo. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "OSSCancellationTokenSource.h"
#import "OSSCancellationTokenRegistration.h"
#import "OSSCancellationToken.h"

@interface OSSCancellationTests : XCTestCase

@end

@implementation OSSCancellationTests

- (void)testCancel {
    OSSCancellationTokenSource *cts = [OSSCancellationTokenSource cancellationTokenSource];
    
    XCTAssertFalse(cts.cancellationRequested, @"Source should not be cancelled");
    XCTAssertFalse(cts.token.cancellationRequested, @"Token should not be cancelled");
    
    [cts cancel];
    
    XCTAssertTrue(cts.cancellationRequested, @"Source should be cancelled");
    XCTAssertTrue(cts.token.cancellationRequested, @"Token should be cancelled");
}

- (void)testCancelMultipleTimes {
    OSSCancellationTokenSource *cts = [OSSCancellationTokenSource cancellationTokenSource];
    XCTAssertFalse(cts.cancellationRequested);
    XCTAssertFalse(cts.token.cancellationRequested);
    
    [cts cancel];
    XCTAssertTrue(cts.cancellationRequested);
    XCTAssertTrue(cts.token.cancellationRequested);
    
    [cts cancel];
    XCTAssertTrue(cts.cancellationRequested);
    XCTAssertTrue(cts.token.cancellationRequested);
}

- (void)testCancellationBlock {
    __block BOOL cancelled = NO;
    
    OSSCancellationTokenSource *cts = [OSSCancellationTokenSource cancellationTokenSource];
    [cts.token registerCancellationObserverWithBlock:^{
        cancelled = YES;
    }];
    
    XCTAssertFalse(cts.cancellationRequested, @"Source should not be cancelled");
    XCTAssertFalse(cts.token.cancellationRequested, @"Token should not be cancelled");
    
    [cts cancel];
    
    XCTAssertTrue(cancelled, @"Source should be cancelled");
}

- (void)testCancellationAfterDelay {
    OSSCancellationTokenSource *cts = [OSSCancellationTokenSource cancellationTokenSource];
    
    XCTAssertFalse(cts.cancellationRequested, @"Source should not be cancelled");
    XCTAssertFalse(cts.token.cancellationRequested, @"Token should not be cancelled");
    
    [cts cancelAfterDelay:200];
    XCTAssertFalse(cts.cancellationRequested, @"Source should be cancelled");
    XCTAssertFalse(cts.token.cancellationRequested, @"Token should be cancelled");
    
    // Spin the run loop for half a second, since `delay` is in milliseconds, not seconds.
    [[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.5]];
    
    XCTAssertTrue(cts.cancellationRequested, @"Source should be cancelled");
    XCTAssertTrue(cts.token.cancellationRequested, @"Token should be cancelled");
}

- (void)testCancellationAfterDelayValidation {
    OSSCancellationTokenSource *cts = [OSSCancellationTokenSource cancellationTokenSource];
    
    XCTAssertFalse(cts.cancellationRequested);
    XCTAssertFalse(cts.token.cancellationRequested);
    
    XCTAssertThrowsSpecificNamed([cts cancelAfterDelay:-2], NSException, NSInvalidArgumentException);
}

- (void)testCancellationAfterZeroDelay {
    OSSCancellationTokenSource *cts = [OSSCancellationTokenSource cancellationTokenSource];
    
    XCTAssertFalse(cts.cancellationRequested);
    XCTAssertFalse(cts.token.cancellationRequested);
    
    [cts cancelAfterDelay:0];
    
    XCTAssertTrue(cts.cancellationRequested);
    XCTAssertTrue(cts.token.cancellationRequested);
}

- (void)testCancellationAfterDelayOnCancelled {
    OSSCancellationTokenSource *cts = [OSSCancellationTokenSource cancellationTokenSource];
    [cts cancel];
    XCTAssertTrue(cts.cancellationRequested);
    XCTAssertTrue(cts.token.cancellationRequested);
    
    [cts cancelAfterDelay:1];
    
    XCTAssertTrue(cts.cancellationRequested);
    XCTAssertTrue(cts.token.cancellationRequested);
}

- (void)testDispose {
    OSSCancellationTokenSource *cts = [OSSCancellationTokenSource cancellationTokenSource];
    [cts dispose];
    XCTAssertThrowsSpecificNamed([cts cancel], NSException, NSInternalInconsistencyException);
    XCTAssertThrowsSpecificNamed(cts.cancellationRequested, NSException, NSInternalInconsistencyException);
    XCTAssertThrowsSpecificNamed(cts.token.cancellationRequested, NSException, NSInternalInconsistencyException);
    
    cts = [OSSCancellationTokenSource cancellationTokenSource];
    [cts cancel];
    XCTAssertTrue(cts.cancellationRequested, @"Source should be cancelled");
    XCTAssertTrue(cts.token.cancellationRequested, @"Token should be cancelled");
    
    [cts dispose];
    XCTAssertThrowsSpecificNamed(cts.cancellationRequested, NSException, NSInternalInconsistencyException);
    XCTAssertThrowsSpecificNamed(cts.token.cancellationRequested, NSException, NSInternalInconsistencyException);
}

- (void)testDisposeMultipleTimes {
    OSSCancellationTokenSource *cts = [OSSCancellationTokenSource cancellationTokenSource];
    [cts dispose];
    XCTAssertNoThrow([cts dispose]);
}

- (void)testDisposeRegistration {
    OSSCancellationTokenSource *cts = [OSSCancellationTokenSource cancellationTokenSource];
    OSSCancellationTokenRegistration *registration = [cts.token registerCancellationObserverWithBlock:^{
        XCTFail();
    }];
    XCTAssertNoThrow([registration dispose]);
    
    [cts cancel];
}

- (void)testDisposeRegistrationMultipleTimes {
    OSSCancellationTokenSource *cts = [OSSCancellationTokenSource cancellationTokenSource];
    OSSCancellationTokenRegistration *registration = [cts.token registerCancellationObserverWithBlock:^{
        XCTFail();
    }];
    XCTAssertNoThrow([registration dispose]);
    XCTAssertNoThrow([registration dispose]);
    
    [cts cancel];
}

- (void)testDisposeRegistrationAfterCancellationToken {
    OSSCancellationTokenSource *cts = [OSSCancellationTokenSource cancellationTokenSource];
    OSSCancellationTokenRegistration *registration = [cts.token registerCancellationObserverWithBlock:^{ }];
    
    [registration dispose];
    [cts dispose];
}

- (void)testDisposeRegistrationBeforeCancellationToken {
    OSSCancellationTokenSource *cts = [OSSCancellationTokenSource cancellationTokenSource];
    OSSCancellationTokenRegistration *registration = [cts.token registerCancellationObserverWithBlock:^{ }];
    
    [cts dispose];
    XCTAssertNoThrow([registration dispose]);
}

@end

File: OSSSwiftDemo/OSSSwiftDemoTests/OSSCancellationTests.swift
//
//  OSSCancellationTests.swift
//  OSSSwiftDemoTests
//
//  Created by huaixu on 2018/1/14.
//  Copyright © 2018年 aliyun. All rights reserved.
//

import XCTest
import AliyunOSSiOS
import AliyunOSSSwiftSDK

class OSSCancellationTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testCancel() {
        let cts = OSSCancellationTokenSource()
        
        XCTAssertFalse(cts.isCancellationRequested, "Source should not be cancelled")
        XCTAssertFalse(cts.token.isCancellationRequested, "Token should not be cancelled")
        cts.cancel()
        
        XCTAssertTrue(cts.isCancellationRequested, "Source should be cancelled")
        XCTAssertTrue(cts.token.isCancellationRequested, "Token should be cancelled")
    }
    
    func testCancelMultipleTimes() {
        let cts = OSSCancellationTokenSource()
        
        XCTAssertFalse(cts.isCancellationRequested)
        XCTAssertFalse(cts.token.isCancellationRequested)
        
        cts.cancel()

        XCTAssertTrue(cts.isCancellationRequested)
        XCTAssertTrue(cts.token.isCancellationRequested)
        
        cts.cancel()
        
        XCTAssertTrue(cts.isCancellationRequested)
        XCTAssertTrue(cts.token.isCancellationRequested)
    }
    
    func testCancellationBlock() {
        var cancelled = false
        let cts = OSSCancellationTokenSource()
        cts.token .registerCancellationObserver {
            cancelled = true
        }
        
        XCTAssertFalse(cts.isCancellationRequested, "Source should not be cancelled")
        XCTAssertFalse(cts.token.isCancellationRequested, "Token should not be cancelled")
        
        cts.cancel()
        
        XCTAssertTrue(cancelled, "Source should be cancelled")
    }
    
    func testCancellationAfterDelay() {
        let cts = OSSCancellationTokenSource()
        XCTAssertFalse(cts.isCancellationRequested, "Source should not be cancelled")
        XCTAssertFalse(cts.token.isCancellationRequested, "Token should not be cancelled")
        
        cts.cancel(afterDelay: 200)
        
        XCTAssertFalse(cts.isCancellationRequested, "Source should be cancelled");
        XCTAssertFalse(cts.token.isCancellationRequested, "Token should be cancelled")
        
        // Spin the run loop for half a second, since `delay` is in milliseconds, not seconds.
        RunLoop.current.run(until: Date.init(timeIntervalSinceNow: 0.5))
        
        XCTAssertTrue(cts.isCancellationRequested, "Source should be cancelled");
        XCTAssertTrue(cts.token.isCancellationRequested, "Token should be cancelled")
    }
    
    func testCancellationAfterDelayValidation() {
        let cts = OSSCancellationTokenSource()
        
        XCTAssertFalse(cts.isCancellationRequested)
        XCTAssertFalse(cts.token.isCancellationRequested)
        
        XCTAssertThrowsError(cts.cancel(afterDelay: -2), NSExceptionName.internalInconsistencyException.rawValue) { (error) in
            XCTAssertNotNil(error)
        }
    }
    
    func testCancellationAfterZeroDelay() {
        let cts = OSSCancellationTokenSource()
        
        XCTAssertFalse(cts.isCancellationRequested)
        XCTAssertFalse(cts.token.isCancellationRequested)
        
        cts.cancel(afterDelay: 0)
        
        XCTAssertTrue(cts.isCancellationRequested);
        XCTAssertTrue(cts.token.isCancellationRequested)
    }
    
    func testCancellationAfterDelayOnCancelled() {
        let cts = OSSCancellationTokenSource()
        
        cts.cancel()
        XCTAssertTrue(cts.isCancellationRequested);
        XCTAssertTrue(cts.token.isCancellationRequested)
        
        cts.cancel(afterDelay: 1)
        
        XCTAssertTrue(cts.isCancellationRequested);
        XCTAssertTrue(cts.token.isCancellationRequested)
    }
    
    func testDispose() {
        var cts = OSSCancellationTokenSource()
        cts.dispose()
        
        XCTAssertThrowsError(cts.cancel(), NSExceptionName.internalInconsistencyException.rawValue) { (error) in
            XCTAssertNotNil(error)
        }
        XCTAssertThrowsError(cts.isCancellationRequested, NSExceptionName.internalInconsistencyException.rawValue) { (error) in
            XCTAssertNotNil(error)
        }
        XCTAssertThrowsError(cts.token.isCancellationRequested, NSExceptionName.internalInconsistencyException.rawValue) { (error) in
            XCTAssertNotNil(error)
        }
        
        cts = OSSCancellationTokenSource()
        cts.cancel()
        
        XCTAssertTrue(cts.isCancellationRequested, "Source should be cancelled")
        XCTAssertTrue(cts.token.isCancellationRequested, "Token should be cancelled")
        
        cts.dispose()
        
        XCTAssertThrowsError(cts.isCancellationRequested, NSExceptionName.internalInconsistencyException.rawValue) { (error) in
            XCTAssertNotNil(error)
        }
        XCTAssertThrowsError(cts.token.isCancellationRequested, NSExceptionName.internalInconsistencyException.rawValue) { (error) in
            XCTAssertNotNil(error)
        }
    }
    
    func testDisposeMultipleTimes() {
        let cts = OSSCancellationTokenSource()
        cts.dispose()
        XCTAssertNoThrow(cts.dispose())
    }
    
    func testDisposeRegistration() {
        let cts = OSSCancellationTokenSource()
        let registration = cts.token.registerCancellationObserver {
            XCTFail()
        }
        XCTAssertNoThrow(registration.dispose())
        cts.cancel()
    }
    
    func testDisposeRegistrationMultipleTimes() {
        let cts = OSSCancellationTokenSource()
        let registration = cts.token.registerCancellationObserver {
            XCTFail()
        }
        
        XCTAssertNoThrow(registration.dispose())
        XCTAssertNoThrow(registration.dispose())
        
        cts.cancel()
    }
    
    func testDisposeRegistrationAfterCancellationToken() {
        let cts = OSSCancellationTokenSource()
        let registration = cts.token.registerCancellationObserver {}
        registration.dispose()
        cts.dispose()
    }
    
    func testDisposeRegistrationBeforeCancellationToken() {
        let cts = OSSCancellationTokenSource()
        let registration = cts.token.registerCancellationObserver {}
        cts.dispose()
        XCTAssertNoThrow(registration.dispose())
    }
}
